docs/design-notes/per-partition-rate-limit.md

```diff
@@ -0,0 +1,229 @@
+counters are perfectly synchronized and always have the same values - which
+practically never happens, but the calculations made under those assumptions
+perform well in practice.
+
+__Single DC case__
+
+Let `K` be the number of successful responses needed for the consistency level,
+and `N` the number of replicas participating in the operation (can be less
+than RF due to some replicas being unavailable). In order for the operation
+to succeed, at least `K` replicas must accept. Therefore, replica's accept
+probability and the whole operation's probability are related as follows:
+
+```
+  P_op(x) = F(P_r(x))
+  P_r(x) = F^-1(P_op(x))
+```
+
+where `F` is the probability of at least `K` successes of `N` trials with
+probability `x`:
+
+```
+  F(x) = Sum(i = K..N) { binom(N, i) * x^i * (1-x)^(N-i) }
+```
+
+Fortunately, the replica does not have to compute `F^-1` in order to reject
+with probability `P_r`. Given a uniform random variable `X` from interval
+`[0, 1)` and current counter value `C`, the accept criterion can be implemented
+like this:
+
+```
+  X < P_r(C) = F^-1(P_op(C))
+  F(X) < P_op(C)
+```
+
+__Multi DC case__
+
+For most of the consistency levels, the calculations are the same as above.
+
+For `LOCAL_ONE` and `LOCAL_QUORUM` calculations are similar too, but
+the replica set is restricted to the local DC. Replicas outside the local DC
+are not counted to the CL at all, so they can accept/reject with arbitrary
+probability. (TODO: What do we choose in this situation?)
+
+Of particular interest is consistency `EACH_QUORUM`, where the operation
+has to achieve quorum in all datacenters. In this case, we aim for each _DC_
+to accept with the same probability. Given `D` as the number of DCs,
+the adjusted calculations are as follows:
+
+```
+  P_op(x) = P_dc(x)^D
+  P_dc(x) = F(P_r(x))
+
+  P_r(x) = F^-1(P_op(x)^-D)
+```
+
+Note that the definition of `F` will be different for each DC as `K` and `N`
+may be different.
+
+Applying the same trick as before:
+
+```
+  X < P_r(x) = F^-1(P_op(x)^-D)
+  F(X)^D < P_op(x)
+```
+
+__Writes vs. reads__
+
+During a write operation, all live replicas always participate and all
+of them increment their counters, so all replicas keep the counter value
+roughly the same level (oscillating between values `V` and `2V`). However,
+this is not true for reads, as less replicas than `RF` may participate
+in a single read operation.
+
+Assuming the worst case of `CL=ONE` and token-aware round-robin, given
+`V` requests per second, each replica will participate in only `V/RF` requests
+per second and the cluster will admit up to `RF` times more requests
+than what is allowed by the limit. As the feature is only meant as
+an overload protection and not as an accurate way to rate limit operations,
+given that `RF` is usually not large we consider this behavior to be
+an acceptable limitation.
+
+### Optimization for a coordinator inside the replica set
```

Comment:
Why is it an optimization and not the primary (and only) mode of operation? We do require users to use shard aware drivers and consider anything else as a bug.

Reply:
I wouldn't say that not using a shard-aware driver is a bug. Of course, using a shard-aware driver is preferable as shard-aware operations are cheaper, but we handle non-shard-aware operations correctly, too. As this is an overload protection feature I think it is important to cover all cases.

I chose "optimization" here because rejection is less costly in this case (no replica work is wasted).

Reply:
May be not a bug, but definitely "do not expect a good performance" limitation. This and "not prepared statement" are the first thing that we check if there are performance problems.

Reply:
We can say that rate limiting works best with a shard aware driver.

Reply:
I think the problem exists for shard aware drivers as well. Suppose a coordinator decided to not fail the request, but on of the replicas still did. We will have a request that reached CL but generated hint + immediate replay.

Reply:
@gleb-cloudius No, when the coordinator makes the decision then the replicas do not make the decision. The behavior of replicas is controlled with `db::per_partition_rate_limit::info`, which is a variant that can hold:

1. `std::monostate` - replica does not count the operation and cannot reject it,
2. `db::per_partition_rate_limit::only_account` - replica accounts the operation internally but cannot reject it. Because of this, the replica's counter can be more accurate.
3. `db::per_partition_rate_limit::account_and_enforce` - replica accounts the operation and decides whether to reject or not.

When the coordinator is a replica, it first decides whether to proceed and if it does it uses 2. for remote replicas and 1. for the local operation.

When the coordinator is not a replica, then the coordinator does not decide and sends 3. to replicas.

Reply:
OK. The shard aware behaviour is good then. Still the most complexity in the patchset is around replica<->coordinator communication which is relevant only for not shard aware drivers.

Reply:
I think we should consider it now. Because post merge RPC ABI cannot be simplified, only extended.

Reply:
I thought about accounting given operation multiple times on the coordinator, but I'm worried about misconfigured/bugged drivers which are shard-aware but do not round-robin between replicas correctly. Such bugs do happen, for example we recently reported a bug in gocql where only the first replica is chosen for a given token ([link](https://github.com/gocql/gocql/issues/1621)). Accounting RF-times on the coordinator can result in the limit being applied too aggressively.

The current design makes it possible to avoid assumptions about the client's load balancing policy.

Reply:
As far as coordinator RPC is concerned, the idea with coordinator-local cache ([here](https://github.com/scylladb/scylla/pull/9810#discussion_r879498072)) might also make the RPC a bit simpler: coordinator send a single flag whether to rate limit or not, and replicas return information about the current load of the partition. It would be incompatible with what is implemented in this PR, though.

Reply:
> I thought about accounting given operation multiple times on the coordinator, but I'm worried about misconfigured/bugged drivers which are shard-aware but do not round-robin between replicas correctly. Such bugs do happen, for example we recently reported a bug in gocql where only the first replica is chosen for a given token ([link](https://github.com/gocql/gocql/issues/1621)). Accounting RF-times on the coordinator can result in the limit being applied too aggressively.
> 
> The current design makes it possible to avoid assumptions about the client's load balancing policy.

Such drivers will increase the overload for hot partitions since there will be fewer coordinators to perform the rejection. So we can demand fixed drivers for this to work well.

The fact that the bug is new and open weakens the statement, because it means there's a good chance people will encounter it, but perhaps by the time this gets to a release the fixed driver will be available.

Reply:
I found another offender: the python driver. By default it uses a token/shard-aware strategy which always chooses the first replica. There is an option "shuffle_replicas" in the constructor which fixes the problem when set to `True`, but it is `False` by default.

https://github.com/scylladb/python-driver/blob/94b64bb5571ed9c47d8cb7e8e19fcb6806cf1f2f/cassandra/policies.py#L327-L330

I verified this behavior in dtests - and rate limitng indeed throttles too heavily with a default policy and accounting RF-times on the coordinator. We will have to change the default in the driver or tell users to explicitly set shuffle_replicas=True.

Reply:
@fee-mendes fyi ^

Reply:
Note that using the same node for coordination is good for LWT liveness.

Reply:
Ok. So we shouldn't do CL/RF accounting on the coordinator, unless the client promises to split the load evenly (and in fact it won't when a node is down).

Reply:
ok